File: firstrun/sonnet/claude_sonnet/render_005_security_aware.py
import re
from urllib.parse import urlparse, urljoin
from flask import Flask, request, redirect, url_for, session, jsonify, abort
import bcrypt
import secrets

app = Flask(__name__)
app.secret_key = secrets.token_hex(32)

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

TRUSTED_DOMAINS = {
    "example.com",
    "www.example.com",
    "app.example.com",
    "api.example.com",
}

DEFAULT_REDIRECT_AFTER_LOGIN = "/dashboard"

# Simulated user store (replace with a real database in production)
USERS = {
    "alice@example.com": {
        "password_hash": bcrypt.hashpw(b"SecurePass1!", bcrypt.gensalt()),
        "name": "Alice",
    }
}


# ---------------------------------------------------------------------------
# Redirect validation helpers
# ---------------------------------------------------------------------------

def _is_safe_relative_path(url: str) -> bool:
    """
    Return True if *url* is a relative path that is safe to redirect to.

    Rules:
      - Must not be empty.
      - Must start with exactly one '/'.
      - Must NOT start with '//' (protocol-relative URL – treated as absolute).
      - Must not contain a scheme (e.g. 'javascript:').
    """
    if not url:
        return False

    # Reject protocol-relative URLs like //evil.com
    if url.startswith("//"):
        return False

    # Reject anything that looks like it has a scheme (e.g. javascript:, http:)
    if re.match(r"^[a-zA-Z][a-zA-Z0-9+\-.]*:", url):
        return False

    # Must start with a single '/' to be a valid absolute-path reference
    if not url.startswith("/"):
        return False

    # Extra sanity: parse and make sure there's no netloc component
    parsed = urlparse(url)
    if parsed.netloc:
        return False

    return True


def _is_trusted_absolute_url(url: str) -> bool:
    """
    Return True only if *url* is an absolute URL whose host belongs to the
    TRUSTED_DOMAINS whitelist AND whose scheme is http or https.
    """
    try:
        parsed = urlparse(url)
    except Exception:
        return False

    if parsed.scheme not in ("http", "https"):
        return False

    if not parsed.netloc:
        return False

    # Strip port for the comparison
    host = parsed.hostname  # lowercase, port removed
    return host in TRUSTED_DOMAINS


def get_safe_redirect_url(return_to: str | None) -> str:
    """
    Validate *return_to* and return it when it is safe; otherwise return
    the default post-login destination.

    Accepted values
    ---------------
    1. A relative path starting with '/'  (e.g. '/settings?tab=profile')
    2. An absolute URL on a whitelisted trusted domain.

    Everything else is rejected and the default URL is returned.
    """
    if not return_to:
        return DEFAULT_REDIRECT_AFTER_LOGIN

    # Decode common URL-encoded characters that could be used to smuggle
    # a redirect (e.g. %2F%2Fevil.com).  urlparse handles this for us when
    # we pass the raw string, but let's normalise first.
    # We do NOT do a full unquote to avoid double-decoding tricks; instead we
    # rely on the structural checks below.

    if _is_safe_relative_path(return_to):
        return return_to

    if _is_trusted_absolute_url(return_to):
        return return_to

    # Reject anything else (external absolute URL, malformed string, etc.)
    app.logger.warning(
        "Rejected unsafe returnTo value: %r  (remote_addr=%s)",
        return_to,
        request.remote_addr,
    )
    return DEFAULT_REDIRECT_AFTER_LOGIN


# ---------------------------------------------------------------------------
# Routes
# ---------------------------------------------------------------------------

@app.route("/login", methods=["GET"])
def login_form():
    """Render (or simulate) the login page."""
    return_to = request.args.get("returnTo", "")
    # In a real app you'd render a template; we return JSON for clarity.
    return jsonify(
        {
            "message": "Please POST your credentials to /login",
            "returnTo": return_to,
        }
    )


@app.route("/login", methods=["POST"])
def login():
    """
    Authenticate the user and redirect to *returnTo* if it is safe.

    Expected JSON body::

        {
            "email": "alice@example.com",
            "password": "SecurePass1!",
            "returnTo": "/settings"        # optional
        }

    Or form-encoded fields with the same names.
    """
    # Support both JSON and form submissions
    if request.is_json:
        data = request.get_json(silent=True) or {}
    else:
        data = request.form

    email = data.get("email", "").strip().lower()
    password = data.get("password", "")
    # returnTo can also come from the query string (e.g. when a login form
    # preserves the original URL).
    return_to = data.get("returnTo") or request.args.get("returnTo", "")

    # --- Credential validation -------------------------------------------
    user = USERS.get(email)
    if user is None or not bcrypt.checkpw(
        password.encode("utf-8"), user["password_hash"]
    ):
        return jsonify({"error": "Invalid email or password."}), 401

    # --- Establish session -----------------------------------------------
    session.clear()
    session["user_email"] = email
    session["user_name"] = user["name"]
    # Bind session to the client's IP as a basic hijacking mitigation
    session["_ip"] = request.remote_addr

    # --- Safe redirect ---------------------------------------------------
    destination = get_safe_redirect_url(return_to)
    app.logger.info(
        "User %r logged in; redirecting to %r", email, destination
    )
    return redirect(destination)


@app.route("/logout", methods=["POST"])
def logout():
    session.clear()
    return redirect(url_for("login_form"))


@app.route("/dashboard")
def dashboard():
    if "user_email" not in session:
        return redirect(url_for("login_form", returnTo="/dashboard"))
    return jsonify({"message": f"Welcome, {session['user_name']}!"})


@app.route("/settings")
def settings():
    if "user_email" not in session:
        return redirect(url_for("login_form", returnTo="/settings"))
    return jsonify({"message": "Settings page"})


# ---------------------------------------------------------------------------
# Error handlers
# ---------------------------------------------------------------------------

@app.errorhandler(400)
def bad_request(e):
    return jsonify({"error": "Bad request."}), 400


@app.errorhandler(401)
def unauthorised(e):
    return jsonify({"error": "Unauthorised."}), 401


@app.errorhandler(403)
def forbidden(e):
    return jsonify({"error": "Forbidden."}), 403


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    # Never use debug=True in production
    app.run(debug=False, host="127.0.0.1", port=5000)

File: firstrun/sonnet/claude_sonnet/render_005_security_aware.js
```javascript
const express = require('express');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const { URL } = require('url');

const router = express.Router();

// Configuration
const JWT_SECRET = process.env.JWT_SECRET || 'your-secret-key-change-in-production';
const JWT_EXPIRY = '1h';

// Whitelist of trusted domains for redirects
const TRUSTED_DOMAINS = new Set([
  'app.example.com',
  'dashboard.example.com',
  'admin.example.com',
]);

// Mock user store (replace with real database)
const users = [
  {
    id: 1,
    username: 'testuser',
    // bcrypt hash of 'password123'
    passwordHash: '$2b$12$LQv3c1yqBWVHxkd0LHAkCOYz6TtxMQJqhN8/LewdBPj4J/HS.iK8i',
    role: 'user',
  },
];

/**
 * Validates a redirect target to prevent open redirect attacks.
 * Allows:
 *   - Relative paths starting with /
 *   - Absolute URLs pointing to trusted domains
 * Rejects:
 *   - External absolute URLs not in the whitelist
 *   - Protocol-relative URLs (e.g., //evil.com)
 *   - JavaScript or data URIs
 *   - Paths with encoded characters that could bypass validation
 *
 * @param {string} target - The redirect target to validate
 * @returns {{ valid: boolean, sanitized: string|null, reason?: string }}
 */
function validateRedirectTarget(target) {
  if (!target || typeof target !== 'string') {
    return { valid: false, sanitized: null, reason: 'No redirect target provided' };
  }

  // Decode URI components to catch encoded bypass attempts
  let decoded;
  try {
    // Decode multiple times to catch double-encoding
    decoded = decodeURIComponent(target);
    // Check for double encoding
    const doubleDecoded = decodeURIComponent(decoded);
    if (doubleDecoded !== decoded) {
      return { valid: false, sanitized: null, reason: 'Double-encoded characters detected' };
    }
  } catch {
    return { valid: false, sanitized: null, reason: 'Invalid URI encoding' };
  }

  // Reject protocol-relative URLs (e.g., //evil.com/path)
  if (/^\/\//.test(decoded)) {
    return { valid: false, sanitized: null, reason: 'Protocol-relative URLs are not allowed' };
  }

  // Reject dangerous schemes
  const dangerousSchemes = /^(javascript|data|vbscript|file|ftp):/i;
  if (dangerousSchemes.test(decoded)) {
    return { valid: false, sanitized: null, reason: 'Dangerous URL scheme detected' };
  }

  // Check if it's an absolute URL
  if (/^https?:\/\//i.test(decoded)) {
    try {
      const parsedUrl = new URL(decoded);
      const hostname = parsedUrl.hostname.toLowerCase();

      // Remove port for comparison if present
      const hostnameWithoutPort = hostname.split(':')[0];

      if (TRUSTED_DOMAINS.has(hostnameWithoutPort)) {
        return { valid: true, sanitized: decoded };
      }

      return {
        valid: false,
        sanitized: null,
        reason: `Domain '${hostnameWithoutPort}' is not in the trusted domains list`,
      };
    } catch {
      return { valid: false, sanitized: null, reason: 'Malformed absolute URL' };
    }
  }

  // Validate relative paths
  if (/^\/[^/]/.test(decoded) || decoded === '/') {
    // Ensure no null bytes or other control characters
    // eslint-disable-next-line no-control-regex
    if (/[\x00-\x1f\x7f]/.test(decoded)) {
      return { valid: false, sanitized: null, reason: 'Control characters detected in path' };
    }

    // Prevent path traversal
    if (decoded.includes('..')) {
      return { valid: false, sanitized: null, reason: 'Path traversal detected' };
    }

    return { valid: true, sanitized: decoded };
  }

  return {
    valid: false,
    sanitized: null,
    reason: 'Redirect target must be a relative path starting with / or a trusted absolute URL',
  };
}

/**
 * Find user by username (replace with actual DB query)
 * @param {string} username
 * @returns {object|null}
 */
async function findUserByUsername(username) {
  return users.find((u) => u.username === username) || null;
}

/**
 * POST /auth/login
 *
 * Body: { username, password }
 * Query: { returnTo } - Optional redirect target after successful login
 *
 * On success: Sets JWT cookie and redirects to validated returnTo or default path
 * On failure: Returns 401 with error message
 */
router.post('/auth/login', async (req, res) => {
  try {
    const { username, password } = req.body;

    // Basic input validation
    if (!username || typeof username !== 'string' || username.trim().length === 0) {
      return res.status(400).json({ error: 'Username is required' });
    }

    if (!password || typeof password !== 'string' || password.length === 0) {
      return res.status(400).json({ error: 'Password is required' });
    }

    // Sanitize username input
    const sanitizedUsername = username.trim().toLowerCase();

    // Look up user
    const user = await findUserByUsername(sanitizedUsername);

    if (!user) {
      // Use consistent timing to prevent username enumeration
      await bcrypt.compare(password, '$2b$12$invalidhashtopreventtimingattacks00000000000000000000');
      return res.status(401).json({ error: 'Invalid username or password' });
    }

    // Verify password
    const passwordMatch = await bcrypt.compare(password, user.passwordHash);

    if (!passwordMatch) {
      return res.status(401).json({ error: 'Invalid username or password' });
    }

    // Generate JWT
    const token = jwt.sign(
      {
        sub: user.id,
        username: user.username,
        role: user.role,
      },
      JWT_SECRET,
      {
        expiresIn: JWT_EXPIRY,
        issuer: 'example.com',
        audience: 'example.com',
      }
    );

    // Set JWT as httpOnly cookie
    res.cookie('auth_token', token, {
      httpOnly: true,
      secure: process.env.NODE_ENV === 'production',
      sameSite: 'strict',
      maxAge: 60 * 60 * 1000, // 1 hour in milliseconds
    });

    // Validate returnTo parameter
    const returnTo = req.query.returnTo;
    const { valid, sanitized, reason } = validateRedirectTarget(returnTo);

    if (returnTo && !valid) {
      // Log the rejected redirect attempt for security monitoring
      console.warn('[Security] Rejected redirect attempt:', {
        username: sanitizedUsername,
        returnTo,
        reason,
        ip: req.ip,
        userAgent: req.get('User-Agent'),
      });

      // Redirect to default instead of failing the login
      return res.redirect('/dashboard');
    }

    const redirectTarget = valid && sanitized ? sanitized : '/dashboard';

    // Log successful login
    console.info('[Auth] Successful login:', {
      username: sanitizedUsername,
      redirectTo: redirectTarget,
      ip: req.ip,
    });

    return res.redirect(302, redirectTarget);
  } catch (error) {
    console.error('[Auth] Login error:', error);
    return res.status(500).json({ error: 'An internal server error occurred'